### ISSUE TITLE
Shortcuts

### ISSUE THREAD

**Issue by auchri** [telegramdesktop/tdesktop#1060]
The goal of this issue is to add the following shortcuts:
## Usage
- [x] Lock telegram with <kbd>CTRL</kbd> + <kbd>L</kbd>
- [ ] Minimize to tray (<kbd>CTRL</kbd> + <kbd>X</kbd> if no text is selected, <kbd>CMD</kbd>  + <kbd>M</kbd> on mac) or open it, if it's already minimized (see #748, global hotkey)
- [x] Completely close telegram (<kbd>CTRL</kbd> + <kbd>Q</kbd>, <kbd>CTRL</kbd> + <kbd>W</kbd> for windows)
- [ ] Open emoji panel when pressing <kbd>TAB</kbd> in text field
- [ ] Insert last sent message into the text field with <kbd>CTRL</kbd> + <kbd>Up</kbd>, on mac <kbd>CMD</kbd> + <kbd>Up</kbd>
- [ ] Mac Fullscreen-Mode <kbd>CMD</kbd> + <kbd>CTRL</kbd> + <kbd>F</kbd> (See #1904)
- [ ] <kbd>Home</kbd> when message editor empty: go to beginning of message history.
- [ ] <kbd>End</kbd> when message editor empty: go to end of message history.
- [ ] <kbd>Shift</kbd> + Click in selection mode (see 4 in #2404)
- [ ] Switch between two last chats (see #3303)
- [ ] Log Out (see #4251)

## Voice messages
- [ ] <kbd>CTRL</kbd> + <kbd>R</kbd> to start recording
- [ ] <kbd>CTRL</kbd> + <kbd>R</kbd> to stop and send recording
- [ ] <kbd>CTRL</kbd> + <kbd>shift (⇧)</kbd> + <kbd>R</kbd> cancel recording

## Emoji

- [ ] <kbd>Tab</kbd> (Single press) - Open Emoji menu
- [ ] <kbd>Tab</kbd> (In Emoji menu _or ultraDoubleTabPressing_) - Open Stickers menu
- [ ] <kbd>Shift</kbd> + <kbd>**←**</kbd> (in Stickers menu, when focused at first sticker pack) - Open GIFs menu
- [ ] <kbd>Shift</kbd> + <kbd>**→**</kbd> (in GIFs menu) - Back to the stickers
- [ ] <kbd>Shift</kbd> + <kbd>**→**</kbd> (in Stickers menu) - Next sticker pack
- [ ] <kbd>Shift</kbd> + <kbd>**←**</kbd> (in Stickers menu) - Previous sticker pack
- [ ] <kbd>**↑**</kbd><kbd>**↓**</kbd> - Move Up / Move Down
- [ ] <kbd>**←**</kbd><kbd>**→**</kbd> - Move Left / Move Right
- [ ] <kbd>SPACE</kbd> - Insert Emoji (no menu closing) / Send Sticker or GIF (and close menu)
- [ ] <kbd>Esc</kbd> - Close menu

(by @DaFri-Nochiterov at #1706)

## Other
- [ ] Native back/forward mouse button support
- [ ] Shortcut to delete conversation
- [ ] Add option to disable shortcuts
- [ ] Shortcut to switch night mode
- [ ] And also add an overview which shortcuts are available (<kbd>CTRL</kbd> + <kbd>F</kbd> to search in current conversation, ...)

(Summary of #3163, #1904, #1706, #1688, #1330, #1244, #1139, #949, #873, #853, #830, #782, #703 and #78)

**Comment by ralesk** [comment]
I guess a related thing would be to make up and down arrows work reliably (sometimes they do nothing, sometimes they scroll the chat history — I'd be perfectly fine if they only ever navigated in the text entry and history scrollback would be pageup/pagedown instead, kinda like in Skype)

**Comment by auchri** [comment]
@ralesk Scrolling through the history with page up/page down works for me

**Comment by ralesk** [comment]
I know, I missed the word "only" there.  My problem is the way the arrows work.

**Comment by Geobert** [comment]
Please make it configurable: some people do not use classical keyboard layout (think dvorak, bépo, colemark etc…)

**Comment by auchri** [comment]
@Geobert thanks, added it to the list :)

**Comment by xander13th** [comment]
[Minimize to tray] and [opening] can be implemented by one hotkey (see #748 for details). Anyway i'm happy to see hotkeys in to-do list :)

**Comment by tostiheld** [comment]
how is the progress on this issue? any way i could help?

**Comment by poxin13** [comment]
:+1: Yes please! Running this on a chromebook within crouton, there is no taskbar in order to close - have to kill the app. Major PITA here :)

**Comment by terryaney** [comment]
Are there any keyboard shortcuts?  Specifically looking for a shortcut to get me to the 'search' field...otherwise, what is the 'keyboard' mechanism to start a chat with a new person?

**Comment by Walkman100** [comment]
<kbd>Ctrl</kbd> + <kbd>F</kbd> searches the current chat, <kbd>Esc</kbd> closes the current chat and focuses the search all box.

**Comment by erisdev** [comment]
would be nice to have `⌘{` and `⌘}` switch between conversations on mac os x since those are standard issue shortcuts for switching between tabs here.

**Comment by Walkman100** [comment]
Woah, <kbd>Ctrl</kbd>+<kbd>Tab</kbd> and <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>Tab</kbd> are really useful, thanks for sharing.
On the contrary, I think those are nicer to press with the left ctrl key, because then I can keep my right hand on the mouse

**Comment by damaru2** [comment]
Why don't also letting the user to change the keys that trigger the shortcuts?

**Comment by erisdev** [comment]
if the shortcuts were in the application's menu like they ought to be then it _would_ be possible to change them through system preferences, just like any other mac app:

<img width="780" alt="screen shot 2016-03-13 at 18 01 47" src="https://cloud.githubusercontent.com/assets/143007/13731677/acd19134-e945-11e5-9564-c82da5c4cffa.png">

**Comment by erisdev** [comment]
@Walkman100 ymmv of course. i'm on a notebook so my trackpad's always in thumb's reach, but i usually keep both hands on the keyboard unless i'm using an art program.

**Comment by nicho90** [comment]
I was missing the `CMD + CTRL + F` Shortcut for Fullscreen-Mode on MacOS, so added a new issue with [Feature-request for fullscreen-mode] #1904 before I saw there is already an issue about shortcuts. Feel free to add this shortcut to the list above and close my issue then.

**Comment by toxsedyshev** [comment]
> Insert last sent message into the text field with CTRL + Up

Like in command line? But what about quick edit?

**Comment by nicho90** [comment]
Thank you @auchri! 👍

**Comment by stek29** [comment]
@toxsedyshev maybe edit if allowed, and just insert if can't edit?

**Comment by toxsedyshev** [comment]
@stek29 In that case, how to insert if edit allowed? :) No, I think there should be two separate commands.

**Comment by LeonelF** [comment]
Ctrl + L should lock the telegram (as it does right now) and minimize to the tray, that would be an awesome feature

**Comment by john-preston** [comment]
@LeonelF Use Ctrl+L Ctrl+W :)

**Comment by LeonelF** [comment]
@john-preston That is what I do, but it would be nice that Ctrl+L do both since you lock it there is no point to keep it on screen, I bet that almost every user that blocks telegram will minimize to tray after blocking it.

**Comment by toxsedyshev** [comment]
@LeonelF I don't :)

**Comment by Walkman100** [comment]
I've recently seen a `shortcuts-custom.json` file in the `tdata` folder, is this file used to customise commands? if so, @LeonelF you could add a rule to Ctrl+L to make it minimise, then see if Ctrl+L locks & minimises Telegram.

**Comment by auchri** [comment]
> I've recently seen a shortcuts-custom.json file in the tdata folder, is this file used to customise commands?

Yes.

> (As a side note: WOOT telegramdesktop is FINALLY a GitHub org!)

:tada:

**Comment by LeonelF** [comment]
@Walkman100 I found the file and the command, 
`{
        "command": "lock_telegram",
        "keys": "ctrl+l"
    },`

But this doesn't seem to be able to aggregate two commands for one key combination, does it?

**Comment by john-preston** [comment]
@LeonelF No aggregation, sorry.

**Comment by LeonelF** [comment]
@john-preston It would be needed to have a command like lock_and_close_telegram so I could assign it to a key combination. Can it be done? :)

**Comment by stek29** [comment]
> Open emoji panel when pressing <kbd>TAB</kbd> in text field

Should I make a PR for this one? I already implemented it [here](/stek29/tdesktop-plus).

**Comment by ralesk** [comment]
I don't think a movement key should open anything.  In normal GUI programs we activate things to open them.

**Comment by Kondra007** [comment]
Maybe `Ctrl+Up` to edit last message and `Ctrl+Down` to reply to last message?

**Comment by toxsedyshev** [comment]
@ralesk Quick edit by Up key is default behavior in all modern messengers like Skype, Slack, etc. Telegram is a messenger, first of all.

**Comment by Kondra007** [comment]
@toxsedyshev This completely ruins scrolling. I mean, in channels `up/down` scrolls, but in normal chats - no. That ruins experience while using Telegam.

**Comment by 0x49D1** [comment]
@Kondra007 came to this thread to say that! Edit message gonna be nice addition(for now we need to right click and Delete the message) and `Ctrl+Up` seems OK hotkey for the action.

**Comment by ralesk** [comment]
@toxsedyshev My reply was about the Tab key, sorry if that wasn't clear.

**Comment by ralesk** [comment]
@Kondra007 @0x49D1 I side with @toxsedyshev on this one though, in other common messengers (Skype, Slack), even IRC clients (Quassel) page up and page down is what scrolls the chat history, up and down navigates inside the edit box, and it's very common that Up on an empty entry loads the previous comment for editing.  This is what people are used to.

Ctrl+Up and Ctrl+Down could be used to access entry history — not editing, creating a new message with the same text; this might not be useful in Telegram, but also a fairly common feature of IRC clients for example.  In slack, the Ctrl-ness can be switched with a setting.

Ctrl+Down for reply is entirely non-mnemonic to me.

**Comment by toxsedyshev** [comment]
@Kondra007 Yes, but who needs one-line scroll using "key up"? You have "page up", mouse wheel, search. That's why other messengers don't scroll by "key up" too.

**Comment by toxsedyshev** [comment]
@ralesk +1 about Ctrl+Down, who needs to reply to the last message with a quote? :)

**Comment by Kondra007** [comment]
@ralesk regarding "up" and "down":  
`Ctrl+Up` to edit, because you **up**date your message which is already in chat.
`Ctrl+Down` to reply, because you add new message below others.

@toxsedyshev "who needs to reply to the last message with a quote?" It's really important in big groups where several dialogues may be present at once. Or just to express emotion on opinion (+1, -1 etc. since Telegram lacks of this by default)

**Comment by ralesk** [comment]
Addendum (again from an accessibility view): *if* the chat history bit will be made focusable, then it would be useful to navigate with up and down there, effectively selecting/highlighting (whatever you want to call it) a particular message, and then pressing a reply shortcut for example (Ctrl+R reply, Ctrl+F forward, anyone? Like in a mail client?) will be actually useful.  Until then, Up/Down doing what they do now is kinda confusing and annoying to be honest

**Comment by ralesk** [comment]
@Kondra007 if it's a reply to the last message, there's not too much need for the reply-ness :) Although I do understand that it fixes any ambiguity.

**Comment by Kondra007** [comment]
@ralesk Unfortunately, when you start replying to currently the last message in big chats, there may appear 1-5 more by the time you finish answering.

**Comment by ralesk** [comment]
@Kondra007 Good point.

**Comment by toxsedyshev** [comment]
@Kondra007 you could just send a message and everybody understand that it's a reply for the last message. If you always reply with quotes, it makes chat unreadable, from my point.

**Comment by toxsedyshev** [comment]
@Kondra007 Yep, good point, but anyway it's unreadable when used for every reply...

**Comment by Kondra007** [comment]
@toxsedyshev No one suggests always making a reply, but still replies are very important in Telegram, since we have supergroups up to 5k.  
That's only my personal opinion, of course.

**Comment by toxsedyshev** [comment]
In that case how to use it? If you want to reply to particular message with a quote but it's already not the last message :) Maybe several presses of Ctrl+Down?

**Comment by Kondra007** [comment]
No, you understood me wrong.  
By the time you **start** replying, that particular message is the last one, but no one guarantees, that after **finishing writing** the message there won't be 1-5 another from other people and on different topic. See the difference?

**Comment by toxsedyshev** [comment]
No, I understand you right. But you said "No one suggests always making a reply". If you don't reply to every message with a quote, one press of Ctrl+Down won't work. If you do - it makes chat unreadable.

**Comment by Kondra007** [comment]
Errrm..what? If you want to quickly start replying to the last message without ever touching your mouse, you can press `Ctrl+Down` and start replying. However if you're sure you'll be understood correctly (see my example above), you may simply write a message without any replies.

What is more, replies allow to quickly scroll through the whole discussion on particular topic reading only relevant messages and automatically skipping irrelevant ones (since you can click on reply and watch that message and so on)

**Comment by toxsedyshev** [comment]
> If you want to quickly start replying to the last message without ever touching your mouse, you can press Ctrl+Down and start replying.

Yep. But if you do it before starting to write your message, you can't cancel it from keyboard (eg. ESC) and it will make chat unreadable (flooded by unnecessary replies) if nobody sent any message while you were typing. I think, the best way is to reply **after** you finished typing your message by several presses of Ctrl+Down to select the message you wanted to reply, if it's already not the last one.

**Comment by Kondra007** [comment]
Or just maybe make it so if you're replying, first `Escape` press removes "reply" from you message and second `Escape` press exits from current chat.

**Comment by toxsedyshev** [comment]
Or both features: Esc to cancel reply and several Ctrl+Down to select what message reply to. When there are more than ~5 messages per minute it's better to use first option, if less - second option.

**Comment by Kondra007** [comment]
That's nice, but I think it should be clearly explained to everyone if this idea goes live.

**Comment by ralesk** [comment]
I still think direct navigation in the history and Ctrl+R for Reply would be a more logical behaviour :)

**Comment by FUNExtreme** [comment]
@auchri It appears  "Insert last sent message into the text field with CTRL + Up, on mac CMD + Up" is in combination with message editing.
I'm currently looking into implementing a few of these, no promises

**Comment by breakmt** [comment]
Could you please do shortcut for hide/open Telegram Window? This hotkey exists in Skype and it is very handy for me.

**Comment by Kondra007** [comment]
@breakmt `Ctrl+M` hides Telegram window.

**Comment by stek29** [comment]
@Kondra007 But doesn't open hidden window.

**Comment by Kondra007** [comment]
@stek29 No, it doesn't. But to open particular window, `Ctrl+M` should be binded as "Open Telegram" **globally**, shouldn't it?

**Comment by ehsaneam** [comment]
how can i change alignment ?

**Comment by stek29** [comment]
@Kondra007 I don't think that App should create system-wide shortcuts by itself. IMO Your OS should allow to bind some keys to specific app.

**Comment by ralesk** [comment]
@stek29 But Windows doesn't offer such facilities (if anything, the only desktop environment with a centralised global shortcut editing system I've seen is KDE), so it's apps that have options to set this up to whatever shortcut you want.  This is idiomatic on Windows, and goes back to at least the WinAmp era.

**Comment by ralesk** [comment]
@ehsaneam There are no paragraph formatting facilities in Telegram as far as I'm aware.

**Comment by stek29** [comment]
@ralesk Then use other system which allows that :)

**Comment by auchri** [comment]
@Kondra007 @ralesk Try [AutoHotkey](https://autohotkey.com/) :)

**Comment by constantx** [comment]
is there a list of all keyboard shorcuts somewhere?

**Comment by axsuul** [comment]
Is anyone experiencing `ctrl+tab` opening up the Emoji window instead of cycling through chats? 

P.S. How do you stylize those keyboard shortcuts in Markdown? :)

**Comment by beschoenen** [comment]
Any update on when the fullscreen shortcut will be added?

**Comment by hobarrera** [comment]
> Any update on when the fullscreen shortcut will be added?

I see little point in a telegram-specific shortcut for this, just use your desktop's native hotkey for this (super+f in my case).

**Comment by beschoenen** [comment]
Well, that doesn't work. At least not on OS X(<kbd>cmd</kbd>+<kbd>shift</kbd>+<kbd>f</kbd>)

**Comment by hobarrera** [comment]
@beschoenen Did you try simply <kbd>cmd</kbd>+<kbd>f</kbd> as well? I'm not sure what the hotkey is, maybe you should check the docs.

But I don't think telegram should have a specific one, just avoid messing up each desktop's hotkey (that *may* be the case on OS X, I'm not sure).

**Comment by CarlosLopezES** [comment]
I think that the shortcut to start and stop recording voice messages would be very useful :)

**Comment by auchri** [comment]
@CarlosLopezES  stop that spam please.

**Comment by ralesk** [comment]
I disagree that <kbd>Ctrl</kbd> + <kbd>W</kbd> would mean “completely close” in Windows parlance, it means “close window” at best, and as such mostly equivalent to <kbd>Alt</kbd> + <kbd>F4</kbd>, and with a program that does stay resident in the tray, it shouldn't do more than hiding the window.  <kbd>Ctrl</kbd> + <kbd>Q</kbd> is the quit shortcut, if anything, but it is rare that programs set it.

**Comment by ralesk** [comment]
I'd like to also point out that <kbd>Ctrl</kbd> + <kbd>X</kbd> to hide the window is a very, very bad idea, certainly confusing to anyone who has never read about it, that something that usually just cuts text (and when nothing is selected, cuts nothing), has such a strange alternative, context sensitive behaviour.  Chiming with my previous command, hide window / minimise to tray should just be <kbd>Ctrl</kbd> + <kbd>W</kbd>.

**Comment by mrjnox** [comment]
Now that #2855 has been closed please merge those two navigation requests into this.

**Comment by ralesk** [comment]
Mentioning #2109 here because it is also about shortcuts, but more specifically the keyboard navigation in chat.

**Comment by MartinWysi** [comment]
anyone know of a mac hotkey/shortcut for desktop app to search repos? cheers.

**Comment by auchri** [comment]
"search repos"?

**Comment by MartinWysi** [comment]
@auchri sure, so in the app on the left hand side there is a list of local repositories, I'm looking for a shortcut/hotkey to search through these without manually clicking on the "filter repositories" search box at the top.

**Comment by auchri** [comment]
You are not talking about Telegram Desktop ^^

**Comment by secretcage** [comment]
i think it would be nice to add ctrl + Arrow Up to editing previous wrote chat . it's getting annoying always when i want to navigate to top.

**Comment by stek29** [comment]
@secretcage was asked for a lot of times. There's a simple workaround btw -- just type something and then use arrows.

**Comment by louwers** [comment]
Can you merge #3163?

- <kbd>Home</kbd> when message editor empty: go to beginning of message history.
- <kbd>End</kbd> when message editor empty: go to end of message history.

---

```
- <kbd>Home</kbd> when message editor empty: go to beginning of message history.
- <kbd>End</kbd> when message editor empty: go to end of message history.
```

**Comment by danger89** [comment]
ctrl+home and ctrl+end keys shall jump to the begin/end of the conversation history.

**Comment by louwers** [comment]
I thought that was a good idea as well @danger89. But then I realized those are already reserved for jumping to the beginning and end of the editor window. As this is default behavior for all editors, it probably shouldn't be meddled with.

<kbd>Home</kbd> and <kbd>End</kbd> are a nice compromise for then the editor is empty, just like <kbd>PgUp</kbd> and <kbd>PgDown</kbd> scroll the messages when the editor is emtpy, and get their default behavior (scrolling the editor) when the editor has contents.

**Comment by danger89** [comment]
@louwers You are right! Just home and end, only when the editor is empty (similar as the page up/down keys).

**Comment by spiritualcoach** [comment]
Any shortcut for calling emoji panel?

**Comment by rhnonose** [comment]
ctrl + shift + tab is not working anymore in OS X
Was it intentionally removed or is it a bug?

**Comment by john-preston** [comment]
@rhnonose Works fine for me on macOS 10.12.4 with Telegram Desktop 1.0.34 alpha. Nothing changed regarding that key shortcut for a couple of years at least.

**Comment by rhnonose** [comment]
@john-preston my telegram says version 2.96.94323 Stable
Is the version disparity so large because of stable/alpha differences?
Checking for update says there's no newer version.
It stopped working a couple of days ago, maybe a week or two.

**Comment by auchri** [comment]
@rhnonose That's a different Telegram App.

**Comment by rhnonose** [comment]
@auchri appears so, thanks

**Comment by john-preston** [comment]
@rhnonose You can try Telegram Desktop from https://desktop.telegram.org

**Comment by WyRe** [comment]
CTRL + L does not work for me.

**Comment by stek29** [comment]
@WyRe What's your system? And please, create a separate issue for this (possible) bug.

**Comment by frizzby** [comment]
How can I hide emoji panel on macosx? It's super annoying.
Already tried every shortcut mentioned here. No result.

**Comment by stek29** [comment]
@frizzby just click on button which opened it?

**Comment by arieljannai** [comment]
@auchri Could you add to the list the shortcut `Shift + Delete` to show the delete confirmation popup with the mark already checked on `Delete for ***` option?
I think it really makes sense with the usage of *delete* and *shift + delete* on various systems

**Comment by BoffinbraiN** [comment]
Is there going to be a menu item that you can use to discover all the shortcut keys in Telegram? Many web apps have a popup that will show you keyboard hints.

**Comment by louwers** [comment]
@BoffinbraiN That suggestion probably deserves it's own issue.

**Comment by BoffinbraiN** [comment]
I'll let @auchri decide on that.
The only keys I know of at the moment are Alt+Up/Down and Ctrl+Q/W/M. If there are several more that already exist, then perhaps it should be added as a separate feature since I'd never find them otherwise.

**Comment by jenkokov** [comment]
@BoffinbraiN you can view all shortcuts avialable now in *Telegram Desktop\tdata\shortcuts-default.json* file, but I definitely agree with you with pop-up idea, and shortcut like **Ctrl+?** for showing it

**Comment by ryukinix** [comment]
Is there any known workaround to lock the voice record at moment?

**Comment by heuristicAL** [comment]
A shortcut like the MacOS app's `cmd + k` would be SUPER useful (allowing us to switch to a specific chat more quickly)

**Comment by amosbird** [comment]
Is there a way to make `alt-f/b` work as `ctrl-left/right` on Linux?

**Comment by prototype99** [comment]
i imagine the easiest way to satisfy a lot of bugs/requests would be to have a hotkey section in settings? and maybe eventually evolve that into having some form of basic scripting to allow custom hotkey commands

**Comment by stek29** [comment]
@prototype99 there actually is one already: see tdata/shortcuts-custom.json

**Comment by prototype99** [comment]
any support for mouse wheel clicks? just figure that would resolve a bug that was filed potentially

**Comment by krupenja** [comment]
Having a shortcut for toggling Night Mode would be amazing.

**Comment by yachen4ever** [comment]
please add the global show/hide interface shortcut key!!!!! friends and i have been waiting for such a long time. It's almost impossible to bear clicking icons on system tray when you are typing or coding.

**Comment by The-DevOps-Guy** [comment]
Is there any progress on the persistent mic record option, its been 5 years now. This is insanity, its 2 lines of code FFS.

**Comment by RememberTheAir** [comment]
@The-DevOps-Guy open a pull request

**Comment by SmartManoj** [comment]
> * Shortcut to delete conversation

 Shortcut to delete multiple conversation also

**Comment by BaseMax** [comment]
Hello,

There is a plan for the `END key`?
@auchri @stek29

Best,
Max

**Comment by shmykelsa** [comment]
Oh God is this still a thing? I would love it

**Comment by andrhua** [comment]
Surprised nobody ever requested vim-style shortcuts, i-l for switching between chat list and current chat, j-k for scrolling, / for searching, etc...

**Comment by Aokromes** [comment]
maybe 

https://doc.qt.io/archives/qq/qq14-actioneditor.html
https://github.com/bjorn/tiled/commit/c1a48ad13b812d6531b372ec8295b19a769257d9

can help someone to create one pull request implementing custom user shortcuts allowing everyone to setup every shortcut they want.

**Comment by damnfella** [comment]
Add this to the checklist:

- [ ] send without sound `CTRL + Enter`
- [ ] schedule the message `CTRL + SHIFT + Enter`
- [ ] send when online `CTRL + ALT + Enter`

**Comment by damnfella** [comment]
> 
> 
> Surprised nobody have ever requested vim-style shortcuts, or navigation at least, h-l for switching between chat list and current chat, j-k for scrolling, / for searching, etc...

`CTRL + F` is just fine for searching. Besides, when the message box is in focus, this would not work, would it? 🤔

**Comment by damnfella** [comment]
just saw a commit for scheduled messages getting `CTRL + ENTER `🤔 
Is scheduling gonna be used more than sending silent messages (messages without sound)?

Please do `CTRL + SHIFT + ENTER` for silent if so. It would be more convenient to not disturb the other person, if you're gonna send a lot a messages.

**Comment by Aokromes** [comment]
what i do when i must send a lot of messages is write the wall of text and send a single message.

**Comment by 23rd** [comment]
> Please do `CTRL + SHIFT + ENTER` for sending silent messages. It would be more convenient to not disturb the other person, if you're gonna send a lot of messages.

@damnfella Now you can open the `tdata/shortcuts-custom.json` file and write it there.
```
[
     {
         "command": "message_silently",
         "keys": "ctrl+shift+enter"
     }
]

```

**Comment by Aokromes** [comment]
@23rd what do you think about allow customize shortcuts with a gui editing that file? :)

**Comment by shmykelsa** [comment]
> > Please do `CTRL + SHIFT + ENTER` for sending silent messages. It would be more convenient to not disturb the other person, if you're gonna send a lot of messages.
> 
> @damnfella Now you can open the `tdata/shortcuts-custom.json` file and write it there.
> 
> ```
> [
>      {
>          "command": "message_silently",
>          "keys": "ctrl+shift+enter"
>      }
> ]
> ```

Is there any documentation on the available commands?

**Comment by 23rd** [comment]
> Is there any documentation on the available commands?

No, there is no documentation.
But you can take a look here.

https://github.com/telegramdesktop/tdesktop/blob/2008b1beeac0977e63bfd8e9f0b7c096ff94cc5b/Telegram/SourceFiles/core/shortcuts.cpp#L57-L85





> what do you think about allow customize shortcuts with a gui editing that file? :)

I think most people use the default hotkey settings. So creating a GUI for editing hotkeys will take a lot of effort with minimal utility.

**Comment by Aokromes** [comment]
well it will allow close all tickets of "i want shortcut for xxx" :P and the one i want customizable shortcuts xd

**Comment by xander13th** [comment]
I don't want to seem intrusive 😏but ~4.5 years have passed..  what about [#748](https://github.com/telegramdesktop/tdesktop/issues/748)?
Is any chance to get it or should I lose a hope and unsubscribe from this topic?)

**Comment by data-man** [comment]
Can be helpful? [qkeysequencewidget](https://github.com/pbek/qkeysequencewidget)

**Comment by AlexWayfer** [comment]
> just saw a commit for scheduled messages getting `CTRL + ENTER `🤔
> Is scheduling gonna be used more than sending silent messages (messages without sound)?
> 
> Please do `CTRL + SHIFT + ENTER` for sending silent messages. It would be more convenient to not disturb the other person, if you're gonna send a lot of messages.

I even thought that it could be like `Ctrl + Enter` triggers menu:

![image](https://user-images.githubusercontent.com/6510020/73121198-bf8ccf80-3f88-11ea-8775-45693127cd46.png)

Then you can navigate through it via arrow keys and press `Enter` again.

**Comment by ralesk** [comment]
Uh, Ctrl+Enter is line break, along with Shift+Enter.  Please don't break people's muscle memory with putting Schedule Message on Ctrl+Enter.

**Comment by damnfella** [comment]
> 
> 
> > Please do `CTRL + SHIFT + ENTER` for sending silent messages. It would be more convenient to not disturb the other person, if you're gonna send a lot of messages.
> 
> @damnfella Now you can open the `tdata/shortcuts-custom.json` file and write it there.
> 
> ```
> [
>      {
>          "command": "message_silently",
>          "keys": "ctrl+shift+enter"
>      }
> ]
> ```

it doesn't seem to be working @23rd

**Comment by damnfella** [comment]
>I even thought that it could be like Ctrl + Enter triggers menu

@AlexWayfer I made a mistake by saying "ctrl+enter". That shortcut is used for new lines. (and has a toggle to change it for sending messages in the chat settings)

`ALT+ENTER` could do it maybe.

**Comment by ralesk** [comment]
Alt+Enter is (was?) commonly used for fullscreen.
I'm thinking maybe it doesn't need to be Enter.

**Comment by dreua** [comment]
@auchri  Maybe add a link to https://github.com/telegramdesktop/tdesktop/wiki/Keyboard-Shortcuts to your top post so that people are informed of what shortcuts we already have.

**Comment by MartinDelille** [comment]
I was previously using the macos version and the *Cmd+K* shortcut was really convenient for switching conversation. What is the equivalent with this client?

**Comment by WilsonNet** [comment]
Voice messaging with keyboard would change my life for better.

**Comment by damnfella** [comment]
> Voice messaging with keyboard would change my life for better.

Ctrl+M

**Comment by yashpalgoyal1304** [comment]
i read the documentation somewhere in the telegram beta chat i think. if it is found, it would be nice to add that in the description here, right??

**Comment by MohammadAlHajj** [comment]
any updates on the issue?

**Comment by KoalaBear84** [comment]
I just miss that you cannot control the popups with the keyboard which is a real usability issue.

**Comment by Aokromes** [comment]
Since i cannot edit 1st post, resume of shortcuts / request of shortcuts:

## Usage
* [x]  Lock telegram with CTRL + L
* [x]  Minimize to tray (CTRL + M) or open it, if it's already minimized (see #748, global hotkey)
* [x]  Completely close telegram (CTRL + Q, CTRL + W for windows)
* [ ]  Open emoji panel when pressing TAB in text field
* [ ]  Insert last sent message into the text field with CTRL + Up, on mac CMD + Up
* [ ]  Mac Fullscreen-Mode CMD + CTRL + F (See #1904)
* [ ]  Home when message editor empty: go to beginning of message history.
* [ ]  End when message editor empty: go to end of message history.
* [ ]  Shift + Click in selection mode (see 4 in #2404)
* [ ]  Switch between two last chats (see #3303)
* [ ]  Log Out (see #4251)
* [ ]  Save Message to saved messages.
* [ ]  Change account to number
* [ ]  Show spelling popup
* [ ]  Select message(s)
* [ ]  Forward message(s)
* [ ]  Delete message(s)
* [ ]  Answer message
 
## Voice messages
* [ ]  CTRL + R to start recording
* [ ]  CTRL + R to stop and send recording
* [ ]  CTRL + shift (⇧) + R cancel recording
 
## Emoji
* [ ]  Tab (Single press) - Open Emoji menu
* [ ]  Tab (In Emoji menu _or ultraDoubleTabPressing_) - Open Stickers menu
* [ ]  Shift + **←** (in Stickers menu, when focused at first sticker pack) - Open GIFs menu
* [ ]  Shift + **→** (in GIFs menu) - Back to the stickers
* [ ]  Shift + **→** (in Stickers menu) - Next sticker pack
* [ ]  Shift + **←** (in Stickers menu) - Previous sticker pack
* [ ]  **↑****↓** - Move Up / Move Down
* [ ]  **←****→** - Move Left / Move Right
* [ ]  SPACE - Insert Emoji (no menu closing) / Send Sticker or GIF (and close menu)
* [ ]  Esc - Close menu

## Multimedia

* [ ] increase volume
* [ ] decrease volume
* [ ] advance video / song
* [ ] rewind video / song 
 
## Other
* [ ]  Native back/forward mouse button support
* [ ]  Shortcut to delete conversation
* [ ]  Add option to disable shortcuts
* [ ]  Shortcut to switch night mode
* [ ]  And also add an overview which shortcuts are available (CTRL + F to search in current conversation, ...)
* [ ]  Boss key

https://github.com/telegramdesktop/tdesktop/blob/dev/Telegram/SourceFiles/core/shortcuts.cpp#L330
Full list.

**Comment by Aokromes** [comment]
https://www.neowin.net/news/powertoys-025-out-with-many-fixes-video-conferencing-mute-utility-coming-next-week it will ship shortcuts editor

**Comment by ralesk** [comment]
> Tab (Single press) - Open Emoji menu
> Tab (In Emoji menu or ultraDoubleTabPressing) - Open Stickers menu

Yikes, that goes against literally every single UI convention since the 80s!  (And has accessibility implications)  Please don't use Tab for this.

**Comment by arieljannai** [comment]
I think that implementing **tab order** would be really useful as long as there are not enough shortcuts, or there's a debate on which shortcuts should be implemented and how. 

* Using <kbd>Tab</kbd> and <kbd>Shift + Tab</kbd> to move between objects and then <kbd>Space</kbd>/<kbd>Enter</kbd> to choose them.
* Using <kbd>F6</kbd> to move between sections/panel, like in the browser.

**Comment by damnfella** [comment]
What are the shortcuts for:
1. "send without sound"
2. "Schedule message"

**Comment by Olegt0rr** [comment]
Really waiting for `CTRL + R to start recording`

**Comment by JosXa** [comment]
`[Question to devs]` - I was just wondering what's generally the roadblock to adding more keybindings to the application, as [the declarative ShortcutManager](https://github.com/telegramdesktop/tdesktop/blob/dev/Telegram/SourceFiles/core/shortcuts.cpp#L330) seems fairly straightforward to configure at first glance, but we're not seeing a lot of change in terms of shortcuts.
Is it more of a UX problem than the actual implementation? Considerations like "if we add this binding what might we break or block for the future" or "we're not sure what is the right binding to use here to have the perfect experience for everyone"?

**Comment by john-preston** [comment]
@JosXa Global shortcuts are actually not very pleasant to deal with, they are listened throughout the app and no way to understand which section should react to the same shortcut at which time. And the overall low priority, I guess. And no clear understanding, which shortcut should do which action.

**Comment by MarioBinder** [comment]
Navigate like Google Reader (r.i.p) with **j** (next message) and **k** (previous message) within a channel would be helpful.

**Comment by mmelnic** [comment]
There is a bug while using a different layout (f.e. Dvorak) from standard English or Russian.
That's the way, how f.e. CTRL+C, CTRL+V, CTRL+X works at VS Code, Chrome, etc. 
But on CTRL+X at Telegram, I have CTRL+Q - that makes it instantly closed and makes the awful experience far Dvorak layout users.

A shortcut system looks like using the value of key but should use VK_ (virtual key codes) to properly work on all customs layouts. 
Tha almost all layouts are on the same place of the keyboard as on the English layout.
[https://docs.microsoft.com/en-us/windows/win32/inputdev/virtual-key-codes](url)

**Comment by timfayz** [comment]
Hello all! Is there any way to bring a context menu just by using shortcuts (say, for a mistyped word to correct it)? Probably 'power' mac users know that in macOS there is no way to activate context menu without touching trackpad (yes, that was a truly tragic realisation). However, some of the applications (like VSCode) provide such functionality within their application using some custom key combination. Is there any way to expect such feature in Telegram somewhere in the future?

**Comment by ralesk** [comment]
On Mac, no idea, on Windows and Linux the Application Menu key on the keyboard provides this functionality, and so does Shift+F10 (Windows and GTK, but not, it seems, Qt/KDE apps).  Looks like on Mac the only thing you can do is use the Mouse Keys accessibility feature (or some scripting) to do the right click from your keyboard.  Unfortunately this will still right click from the mouse position, rather than from the position of the keyboard focus, like it would do on the other systems.

**Comment by dreua** [comment]
@timfayz @ralesk sound like a bug in macOS to me rather than a bug in Telegram.

**Comment by ralesk** [comment]
Or more like a lack of feature :D  Not much stopping from Telegram to go the Microsoft/GNOME way and bind context menu to Shift+F10 too though.

**Comment by dreua** [comment]
True, I think Shift+F10 should be added to @auchri 's list. (Thanks for teaching me that, never used it before. Didn't need it either as my keyboard has an "application menu" key.

**Comment by timfayz** [comment]
That would be really of a great help for these deprived mac users!

**Comment by damianetienne** [comment]
> What are the shortcuts for:
> 
> 1. "Send without sound"
> 2. "Schedule message"

I really need these

**Comment by Partizanin** [comment]
> > What are the shortcuts for:
> > 
> > 1. "Send without sound"
> > 2. "Schedule message"
> 
> I really need these

it is already exist 
https://github.com/telegramdesktop/tdesktop/issues/7295#issuecomment-657513221

**Comment by iasj** [comment]
This feature would be perfect on a complete control over the functions via rc files somewhere. I dream of the day a could control Telegram with vim-like keys, and also a normal and insert modes would kill anybody.

**Comment by driesvints** [comment]
I'm dying to get support for "Mac Fullscreen-Mode CMD + CTRL + F" 😅 

Is there any way I can help with getting support for this?

**Comment by K0media** [comment]
> I'm dying to get support for "Mac Fullscreen-Mode CMD + CTRL + F" 😅
> 
> Is there any way I can help with getting support for this?

Since Mac users have been asking for <kbd>Ctrl</kbd>/<kbd>Cmd</kbd>+<kbd>Shift</kbd>+<kbd>F</kbd> for full-screen mode, I'd like to suggest **something to jump directly to global search.**

If it can't be <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>F</kbd> to jump to global search... What about <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>L</kbd> or <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>G</kbd>? Or even <kbd>Ctrl</kbd>+<kbd>G</kbd> (that one isn't being used yet, luckily).

Anyone else who agrees?

Related issues:
1. https://github.com/telegramdesktop/tdesktop/issues/2159
2. https://github.com/telegramdesktop/tdesktop/issues/1904

**Comment by artiomqa** [comment]
It would be just great to have an option to switch on/off global shortkeys for both start, and stop & send voice recording message. Having Ctrl+R when the window is active is so much better than holding LMB, but it would be a killer feature I guess IMHO.

**Comment by evanpurkhiser** [comment]
I see we nnow have a `cmd+k` to switch channels! That's great!

Not sure if this is the right place for this request, but it would be great if it filtered chats with unread messages to the top (similar to slack). And also if it didn't take 2 escape presses to exit :)

**Comment by aliazani** [comment]
I think it's better to have a specific panel in settings to customize shortcuts and have a checkbox to tick use vim navigation or not.
This way anyone can change the shortcuts easily and it's not necessary to stick to defaults.

**Comment by pe5ha** [comment]
side mouse back/forward buttons for moving to previously opened chats is MUST HAVE

**Comment by pbaykalov** [comment]
Tdesktop stopped closing media by mouse' back button and this seems like an issue to me. Would be nice if there was a comman for that to at least add it as a custom shortcut.

**Comment by bushnerd** [comment]
> [Minimize to tray] and [opening] can be implemented by one hotkey (see #748 for details). Anyway i'm happy to see hotkeys in to-do list :)

But I believe that native support for this feature is the best. AutoHotkey is like starting another process to monitor the keys all the time.

**Comment by Un2mercey** [comment]
Pls add opportunity to remove `Record audio/video message`
Have a bad flow on working chat 😢

**Comment by j4hangir** [comment]
⌘+R used to work before but in the latest version has stopped working suddenly here.

**Comment by N1ckn1ght** [comment]
Make an option to DISABLE quit shortcut, PLEASE


- fellow windows user

**Comment by ChloeSheasby** [comment]
Is there a way to make the Command-Q shortcut not force you to hold it to quit?

**Comment by john-preston** [comment]
@ChloeSheasby There is a option in Settings for that: Settings > Advanced > System integration > Show warning before quitting with CmdQ.

**Comment by ChloeSheasby** [comment]
@john-preston thank you! I could not find that for the life of me.

**Comment by recolic** [comment]
before that, please allow user to disable these fancy keybindings.
It's preventing me from switching my input method, and also conflict with DE shortcuts. It's so annoying

**Comment by yashikada** [comment]
Is there a shortcut for switch media while the media viewer is in full screen mode and playing a video?
All combinations with left/right arrow seek video and it's ok, but none change media, I need switch to normal screen with CTRL+F, switch media with left arrow and give another CTRL+F for full screen, is boring.

**Comment by iiasceriwebmaster** [comment]
I already enjoyed main ones ctrl l ctrl w, others would be nice but these are the basics and they work thx tg team

**Comment by AlexWayfer** [comment]
I really need for a toggle hotkey to mute in voice calls.

**Comment by laraws** [comment]
need these shortcuts, thx.

 CTRL + R to start recording
 CTRL + R to stop and send recording
 CTRL + shift (⇧) + R cancel recording

**Comment by Parsifa1** [comment]
Is it possible to add a shortcut key to quickly open the main panel?

**Comment by NikKovIos** [comment]
Please, add code shortcut, in order not to write ``` every time.

**Comment by dreua** [comment]
Unsubscribing, the request to progress ratio is so not worth it :roll_eyes:

**Comment by sophronesis** [comment]
would be nice to have shortcuts for reactions to posts
currently doing it with https://github.com/moverest/wl-kbptr

**Comment by noiob** [comment]
the new control-tab chat switcher is horrible and much worse than the old behavior (moving up/down in the chat list, which is already sorted by recency)

**Comment by KEOVICHE** [comment]
1000000

**Comment by RememberTheAir** [comment]
@noiob not the same thing, the recent chats switcher switches across chats that you've recently opened, not the chats that have been active recently. If you have an active chats list (eg more groups active at the same time), ctrl-tab skips the chats you're not interested in and doesn't make you mark as read messages from chats that you don't want to read yet

**Comment by pbaykalov** [comment]
@RememberTheAir sure but pgdn is damn far away from Tab if you ask me

**Comment by pbaykalov** [comment]
@noiob it's strange decision considering there's no UI to customise it but you can still use Alt+Up/Down for that

**Comment by john-preston** [comment]
@pbaykalov No ui to customize what? Shortcuts can be customized in settings > chat settings.

**Comment by pbaykalov** [comment]
@john-preston specifically, there is the dialog, but Ctrl+Tab is not mentioned there.

**Comment by john-preston** [comment]
@pbaykalov You can set Ctrl+Tab / Ctrl+Shift+Tab to the same actions that have Alt+Up / Alt+Down and they'll be used for that instead of the default.
